### PR TITLE
fix(auth): restore Application Default Credentials support to idtoken

### PR DESCRIPTION
**Pull request by quartzmo**

### CODE REVIEW

File: auth/credentials/idtoken/idtoken.go [hunk 1]
```diff
@@ -113,13 +113,20 @@ func NewCredentials(opts *Options) (*auth.Credentials, error) {
...
+	b := opts.jsonBytes()
+	if b == nil && metadata.OnGCE() {
 		return computeCredentials(opts)
+	} else if b == nil {
+		creds, err := credentials.DetectDefault(&credentials.DetectOptions{
+			Audience: opts.Audience,
+		})
+		if err != nil {
+			return nil, err
+		}
```
Comment (quartzmo) [inline auth/credentials/idtoken/idtoken.go:125]: Note that in the case of ADC and impersonated or external account credential types, the `creds` created here are discarded, then re-created in a new call to `DetectDefault` (with additional `DetectOptions` args) in `credsFromBytes`.

File: auth/credentials/idtoken/idtoken.go [hunk 2]
```diff
@@ -113,13 +113,20 @@ func NewCredentials(opts *Options) (*auth.Credentials, error) {
...
-	if b := opts.jsonBytes(); b != nil {
-		return credsFromBytes(b, opts)
-	}
-	if metadata.OnGCE() {
+	b := opts.jsonBytes()
+	if b == nil && metadata.OnGCE() {
 		return computeCredentials(opts)
+	} else if b == nil {
+		creds, err := credentials.DetectDefault(&credentials.DetectOptions{
+			Audience: opts.Audience,
```
Comment (codyoss) [inline auth/credentials/idtoken/idtoken.go:121]: instead of calling the jsonBytes method should we just call DetectDefault and pass in the opts here.

Comment (codyoss) [inline auth/credentials/idtoken/idtoken.go:121]: This aud is for the idtoken, not the creds. This should be some default scopes

File: auth/credentials/idtoken/idtoken.go (hunk 1 above)
Comment (codyoss) [reply auth/credentials/idtoken/idtoken.go:125]: I think those opts could just be all passed above.

**Review by codyoss — CHANGES_REQUESTED** [review]

File: auth/credentials/idtoken/idtoken.go (hunk 2 above)
Comment (quartzmo) [reply auth/credentials/idtoken/idtoken.go:121]: `jsonBytes` is called in order to determine if we should skip `DetectDefault` and go immediately into the `idtoken.computeCredentials` path. Thus the precedence is as follows. Let me know if it's wrong.

1. CredentialsFile or CredentialsJSON is present
2. OnGCE == true
3. DetectDefault.

Comment (quartzmo) [reply auth/credentials/idtoken/idtoken.go:121]: Question: If we just call DetectDefault and pass in the opts here, what do we do with the resulting `creds`? If we simply call `credsFromBytes(creds.JSON(), opts)` with it, we are mostly throwing away the opts passed here.

File: auth/credentials/idtoken/idtoken.go (hunk 1 above)
Comment (quartzmo) [reply auth/credentials/idtoken/idtoken.go:125]: See questions above.

Comment (quartzmo) [reply auth/credentials/idtoken/idtoken.go]: I refactored to use `DetectDefault` in all non-compute cases, which simplifies the code a bit.

File: auth/credentials/idtoken/idtoken.go (hunk 2 above)
Comment (quartzmo) [reply auth/credentials/idtoken/idtoken.go]: Fixed, thanks!

Comment (quartzmo) [reply auth/credentials/idtoken/idtoken.go]: I refactored to use `DetectDefault` in all non-compute cases, which simplifies the code a bit.

File: auth/credentials/idtoken/idtoken.go [hunk 3]
```diff
@@ -113,13 +113,18 @@ func NewCredentials(opts *Options) (*auth.Credentials, error) {
...
 	}
-	return nil, fmt.Errorf("idtoken: couldn't find any credentials")
+	creds, err := credentials.DetectDefault(&credentials.DetectOptions{
+		Scopes:          defaultScopes,
+		CredentialsJSON: b,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return credsFromBytes(creds.JSON(), opts)
```
Comment (codyoss) [inline auth/credentials/idtoken/idtoken.go:127]: Can this be refactored a little more to take the whole creds now, that was you don't have to call detect default again later for one of the flows.

Comment (quartzmo) [reply auth/credentials/idtoken/idtoken.go]: Updated, PTAL!

**Review by codyoss — APPROVED** [review]

**Pull request merged**